### PR TITLE
Add "excluding" optional parameter to TargetPlatformVariant to communicate cases where test should be ran everywhere but specific platforms

### PR DESCRIPTION
**Pull request by antholeole**
Added an optional "excluding" flag for when the user wants to run a test on all platforms *but* some specific platforms. I also changed some occurances where the writer of the test could have used "excluding" but didn't have it available. 

I think in a lot of the cases, the writer of the test listed all the platforms besides the apple platforms - what they probably meant was that they wanted the test run on all the platforms besides the apple ones, but resorted to just manually writing them all out. In the case that the flutter team decides to add more platforms (are there even any more!), this case would be considered broken since a test that just hard codes "all platforms but apple" would have not listed the new platforms (I actually think there are some tests that are being skipped on windows that shouldn't be?). By using "excluding" the author can communicate that what they mean is they want to run that test on all platforms besides the ones listed.

It seems like in a ton of places in the code, we're hard coding in all the other platforms besides one or two - I'd like to replace them with excluding, but their intention is now lost with time: perhaps they _actually_ meant they wanted to only run on those 4 platforms, so the substitution cannot be made. 

I'd recommend looking at this commit by commit, since the first commit is the change and the test, and the second commit just changes some previous tests where this would apply.

*List which issues are fixed by this PR. You must list at least one issue.*

No issues 😬 should I make one?

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 1]
```diff
@@ -254,8 +254,11 @@ class TargetPlatformVariant extends TestVariant<TargetPlatform> {
   const TargetPlatformVariant(this.values);
 
   /// Creates a [TargetPlatformVariant] that tests all values from
-  /// the [TargetPlatform] enum.
-  TargetPlatformVariant.all() : values = TargetPlatform.values.toSet();
+  /// the [TargetPlatform] enum. If [excluding] is provided, will test all platforms
+  /// except those in [excluding].
+  TargetPlatformVariant.all({  
+    Set<TargetPlatform> excluding = const <TargetPlatform>{}
```
Comment (justinmc) [inline packages/flutter_test/lib/src/widget_tester.dart]: Comma at the end here.

File: packages/flutter_test/test/widget_tester_test.dart [hunk 2]
```diff
@@ -720,13 +720,26 @@ void main() {
...
+          expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
+        } else {
+          numberOfVariationsRun += 1;
+        }
+      }, variant: TargetPlatformVariant.all());
+
+      const Set<TargetPlatform> excludePlatforms = <TargetPlatform>{ TargetPlatform.android, TargetPlatform.linux };
+      testWidgets('TargetPlatformVariant.all with excluding runs an all variants except those provided in excluding', (WidgetTester tester) async {
+        if (debugDefaultTargetPlatformOverride == null) {
+          expect(numberOfVariationsRun, equals(TargetPlatform.values.length - excludePlatforms.length));
```
Comment (justinmc) [inline packages/flutter_test/test/widget_tester_test.dart:735]: Nit: Maybe also expect that the current target platform is not in excludePlatforms?

File: packages/flutter_test/test/widget_tester_test.dart [hunk 3]
```diff
@@ -720,13 +720,26 @@ void main() {
...
+        } else {
+          numberOfVariationsRun += 1;
+        }
+      }, variant: TargetPlatformVariant.all(excluding: excludePlatforms));
+    });
+
+
```
Comment (justinmc) [inline packages/flutter_test/test/widget_tester_test.dart:742]: Remove these two extra newlines.

**Review by justinmc — COMMENTED** [review]
I really like this and have found myself wishing it existed before.  I think the possibility of Flutter adding a new platform in the future is a strong argument in favor of adding this.

This LGTM but I'm going to hold off on approving to let others chime in that might have more thoughts about expanding this API or not.

File: packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart [hunk 4]
```diff
@@ -47,7 +47,7 @@ void main() {
 
     expect(topBeforeScroll.dy, equals(0.0));
     expect(topAfterScroll.dy, equals(0.0));
-  }, variant: TargetPlatformVariant(TargetPlatform.values.where((TargetPlatform value) => value != TargetPlatform.fuchsia).toSet()));
+  }, variant: TargetPlatformVariant.all(excluding: <TargetPlatform>{ TargetPlatform.fuchsia }));
```
Comment (Carlos123z) [inline packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart:50]:
```suggestion
-   }, variant: TargetPlatformVariant.all(excluding: <TargetPlatform>{ TargetPlatform.fuchsia }));1
```

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 1 above)
Comment (Carlos123z) [reply packages/flutter_test/lib/src/widget_tester.dart]: No

File: packages/flutter_test/test/widget_tester_test.dart [hunk 5]
```diff
@@ -720,13 +720,26 @@ void main() {
...
-    testWidgets('TargetPlatformVariant.all tests run all variants', (WidgetTester tester) async {
-      if (debugDefaultTargetPlatformOverride == null) {
-        expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
-      } else {
-        numberOfVariationsRun += 1;
-      }
-    }, variant: TargetPlatformVariant.all());
+    group('all', () {
+      testWidgets('TargetPlatformVariant.all tests run all variants', (WidgetTester tester) async {
+        if (debugDefaultTargetPlatformOverride == null) {
```
Comment (Carlos123z) [inline packages/flutter_test/test/widget_tester_test.dart:725]: - -

File: packages/flutter_test/test/widget_tester_test.dart [hunk 6]
```diff
@@ -720,13 +720,26 @@ void main() {
       expect(defaultTargetPlatform, equals(TargetPlatform.iOS));
     }, variant: TargetPlatformVariant.only(TargetPlatform.iOS));
 
-    testWidgets('TargetPlatformVariant.all tests run all variants', (WidgetTester tester) async {
```
Comment (Carlos123z) [inline packages/flutter_test/test/widget_tester_test.dart:723]: -

Comment (Carlos123z) [reply packages/flutter_test/test/widget_tester_test.dart:723]: L mmm

File: packages/flutter_test/test/widget_tester_test.dart [hunk 7]
```diff
@@ -720,13 +720,26 @@ void main() {
...
-      if (debugDefaultTargetPlatformOverride == null) {
-        expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
-      } else {
-        numberOfVariationsRun += 1;
-      }
-    }, variant: TargetPlatformVariant.all());
+    group('all', () {
+      testWidgets('TargetPlatformVariant.all tests run all variants', (WidgetTester tester) async {
+        if (debugDefaultTargetPlatformOverride == null) {
+          expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
```
Comment (Carlos123z) [inline packages/flutter_test/test/widget_tester_test.dart:726]: - 
```suggestion
          expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
```

File: packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart (hunk 4 above)
Comment (Carlos123z) [inline packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart:50]: - [ ]

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 1 above)
Comment (Carlos123z) [reply packages/flutter_test/lib/src/widget_tester.dart]: Yes

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 8]
```diff
@@ -254,8 +254,11 @@ class TargetPlatformVariant extends TestVariant<TargetPlatform> {
   const TargetPlatformVariant(this.values);
 
   /// Creates a [TargetPlatformVariant] that tests all values from
-  /// the [TargetPlatform] enum.
-  TargetPlatformVariant.all() : values = TargetPlatform.values.toSet();
+  /// the [TargetPlatform] enum. If [excluding] is provided, will test all platforms
+  /// except those in [excluding].
+  TargetPlatformVariant.all({
+    Set<TargetPlatform> excluding = const <TargetPlatform>{}
+  }) : values = TargetPlatform.values.toSet()..removeAll(excluding);
```
Comment (Carlos123z) [inline packages/flutter_test/lib/src/widget_tester.dart:261]: Yes

File: packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart [hunk 9]
```diff
@@ -47,7 +47,7 @@ void main() {
 
     expect(topBeforeScroll.dy, equals(0.0));
     expect(topAfterScroll.dy, equals(0.0));
-  }, variant: TargetPlatformVariant(TargetPlatform.values.where((TargetPlatform value) => value != TargetPlatform.fuchsia).toSet()));
```
Comment (Carlos123z) [inline packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart:50]: - [ ]

Comment (Carlos123z) [inline packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart:50]: - ```

```

**Review by Carlos123z — COMMENTED** [review]
@

File: packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart (hunk 9 above)
Comment (antholeole) [reply packages/flutter/test/material/flexible_space_bar_collapse_mode_test.dart:50]: Come on man be better than that

File: packages/flutter_test/test/widget_tester_test.dart [hunk 10]
```diff
@@ -720,13 +720,26 @@ void main() {
...
+        } else {
+          numberOfVariationsRun += 1;
+        }
+      }, variant: TargetPlatformVariant.all());
+
+      const Set<TargetPlatform> excludePlatforms = <TargetPlatform>{ TargetPlatform.android, TargetPlatform.linux };
+      testWidgets('TargetPlatformVariant.all with excluding runs an all variants except those provided in excluding', (WidgetTester tester) async {
```
Comment (gspencergoog) [inline packages/flutter_test/test/widget_tester_test.dart:733]:
```suggestion
      testWidgets('TargetPlatformVariant.all, excluding runs an all variants except those provided in excluding', (WidgetTester tester) async {
```

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Great idea.

**Pull request merged**